lib/membrane_h264_plugin/parser.ex

```diff
@@ -139,63 +180,100 @@ defmodule Membrane.H264.Parser do
-        profile: nil,
-        previous_buffer_timestamps: nil,
-        frame_prefix: <<>>,
-        parameter_sets_present?: byte_size(sps) > 0 or byte_size(pps) > 0,
-        cached_sps: %{},
-        cached_pps: %{}
-      }
-      |> Map.merge(Map.from_struct(opts))
+    state = %{
+      nalu_splitter: nil,
+      nalu_parser: nil,
+      au_splitter: AUSplitter.new(),
+      au_timestamp_generator: au_timestamp_generator,
+      mode: nil,
+      profile: nil,
+      previous_buffer_timestamps: nil,
+      output_alignment: opts.output_alignment,
+      frame_prefix: <<>>,
+      skip_until_keyframe: opts.skip_until_keyframe,
+      repeat_parameter_sets: opts.repeat_parameter_sets,
+      cached_spss: %{},
+      cached_ppss: %{},
+      initial_spss: initial_parameters_to_list(opts.sps),
+      initial_ppss: initial_parameters_to_list(opts.pps),
+      input_stream_structure: nil,
+      output_stream_structure: output_stream_structure
+    }
 
     {[], state}
   end
 
   @impl true
-  def handle_stream_format(:input, stream_format, _ctx, state) do
-    state =
+  def handle_stream_format(:input, stream_format, ctx, state) do
+    {alignment, input_raw_stream_structure} =
       case stream_format do
-        %H264{alignment: alignment} ->
-          mode =
-            case alignment do
-              :nalu -> :nalu_aligned
-              :au -> :au_aligned
-            end
+        %RemoteStream{type: :bytestream} ->
+          {:bytestream, :annexb}
 
-          %{state | mode: mode}
+        %H264{alignment: alignment, stream_structure: stream_structure} ->
+          {alignment, stream_structure}
+      end
 
-        %RemoteStream{type: :bytestream} ->
-          %{state | mode: :bytestream}
+    is_first_received_stream_format = is_nil(ctx.pads.input.stream_format)
 
-        %H264.RemoteStream{alignment: alignment, decoder_configuration_record: dcr} ->
-          mode =
-            case alignment do
-              :nalu -> :nalu_aligned
-              :au -> :au_aligned
-            end
+    mode = get_mode_from_alignment(alignment)
 
-          %{state | mode: mode, frame_prefix: get_frame_prefix!(dcr, state)}
+    state =
+      cond do
+        is_first_received_stream_format ->
+          input_stream_structure = parse_raw_stream_structure(input_raw_stream_structure)
+
+          output_stream_structure =
+            if is_nil(state.output_stream_structure),
+              do: input_stream_structure,
+              else: state.output_stream_structure
+
+          %{
+            state
+            | mode: mode,
+              nalu_splitter: NALuSplitter.new(input_stream_structure),
+              nalu_parser: NALuParser.new(input_stream_structure),
+              input_stream_structure: input_stream_structure,
+              output_stream_structure: output_stream_structure
+          }
+
+        not is_input_stream_structure_change_allowed?(
+          input_raw_stream_structure,
+          state.input_stream_structure
```

Comment:
Wouldn't it be better to compare the "parsed" stream_structure representation?
```suggestion
        not is_input_stream_structure_change_allowed?(
          input_stream_structure,
          state.input_stream_structure
```

Reply:
it's the same and we are not parsing the `input_raw_stream_structure` in non-first `handle_stream_format`, where this check is invoked